### ISSUE TITLE
[LSP] Order messages in the LSP server

### ISSUE THREAD

**Issue by dibarbet** [dotnet/roslyn#46051]
JsonRPC doesn't necessarily preserve order of responses.  We need to ensure that messages are ordered properly (especially once we implement text document change notifications) so that we do not process things out of order.  We shouldn't serialize all messages as the LSP spec allows us to process some messages in parallel.

TODO:
* [x] Introduce a RequestContext to hold all of the parameters that a handler needs (#46557)
* [x] Create a RequestExecutionQueue to run requests in order and block as necessary (#46660)
* [x] Handlers to provide a way to indicate which document they need to operate on, which is passed back to them (#46896)
* [x] Remove SolutionProvider from AbstractRequestHandler to force new handlers to use the RequestContext.Solution (#47022)
* [x] `mutatesSolutionState` parameter on `ExportLspMethodAttribute` to be non-optional (#47022)
  * and cleanup: https://github.com/dotnet/roslyn/pull/46660#discussion_r468172882
* [x] Queue lifetime should be per LSP-server, not shared via MEF, and problems should close the connection (#47107)
  * see: https://github.com/dotnet/roslyn/pull/46660#discussion_r468227975
  * and: https://github.com/dotnet/roslyn/pull/46660#discussion_r468229986
  * and: https://github.com/dotnet/roslyn/pull/46660#discussion_r471669025
* [x] Does StreamJsonRpc guarantee that message B won't start processing until message A at least returns a task from its handler?
  * ie, can a message be added to the async queue in the wrong order, if a previous message gets unlucky threading
    * Yes, after v2.6 StreamJsonRpc ensures ordering by default until the first await: https://github.com/microsoft/vs-streamjsonrpc/blob/master/doc/resiliency.md#default-ordering-and-concurrency-behavior
* [ ] Does/Can GetTextDocumentWithContextHandler use the right solution snapshots?
* Actual solution merging will come with #45427

**Comment by CyrusNajmabadi** [comment]
> Jsonrpc doesn't necessarily preserve order

i litearlly do not understand how one could possibly build any system that uses jsonrpc if it cannot give ordering.

**Comment by davidwengier** [comment]
Random notes:
* Order of messages _sent_ is preserved, but responses come back whenever they're available
* There is a mechanism to provide a SynchronizationContext on a connection to force an order
* This won't work for us as some methods can run in parallel (eg, FAR and GTD)
* Need a way to flag methods as `read` or `write`
* When a response comes back, queue it up for processing
* Any message sent _after_ a write message has to wait for that write response to be processed
* If there aren't any write messages waiting we can just process as responses come in

Razor has this same problem, and solves it using a helper from OmniSharp:
https://github.com/dotnet/aspnetcore-tooling/blob/06625ac6b98599f7fe1902bbe1ef59ce165469ba/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorLanguageServer.cs#L64

**Comment by davidwengier** [comment]
We can get ordering of messages for free by using NonConcurrentSynchronizationContext from vs-threading. We can even use it to enable parallel reads that block for writes simply by putting an `await TaskScheduler.Default;` at the start of any read method. **BUT** as soon as any write method does an `await` of any kind, it unblocks the rest of the queue, so it might not be a viable solution. This is true even with the `sticky` parameter set to true, as that only ensures that a write and a read won't run concurrently, it doesn't ensure a write will run to completion.

**Comment by davidwengier** [comment]
So I've looked at a few different things for ordering messages and my head is somewhat swimming, so i'm just gonna brain-dump a bit here:

vs-threading's [NonConcurrentSynchronizationContext](https://github.com/microsoft/vs-threading/blob/master/src/Microsoft.VisualStudio.Threading/NonConcurrentSynchronizationContext.cs)
* Readily available to use
* Breaks on the first await (not that I can _really_ see how, I understand the logic)

OmniSharp's [PriorityScheduler](https://github.com/OmniSharp/csharp-language-server-protocol/blob/a8e44c27cb4319bd0e6ffc77df75f9359752b563/src/JsonRpc/ProcessScheduler.cs#L78-L86)
* Now rewritten using System.Reactive, so have to go off an old commit or its hard to track
* Heavily tied to the rest of csharp-language-server-client which is an alternative JsonRpc essentially, so its not exactly a simple "port to Roslyn"
* It wired up in [InputHandler](https://github.com/OmniSharp/csharp-language-server-protocol/blob/a8e44c27cb4319bd0e6ffc77df75f9359752b563/src/JsonRpc/InputHandler.cs) though, so _maybe_ we can provide our own [MessageHandler](https://github.com/microsoft/vs-streamjsonrpc/blob/master/src/StreamJsonRpc/HeaderDelimitedMessageHandler.cs) to do similar?
* Uses a dedicated single thread for all requests which seems bad

Blazor's [RendererSynchronizationContext](https://github.com/dotnet/aspnetcore/blob/master/src/Components/Components/src/Rendering/RendererSynchronizationContext.cs)
* Similar to NonConcurrentSynchronizationContext
* Maybe doesn't break on the first await, not that I have a high confidence in saying that
* I think could be a good fit with the dispatcher that goes with it being our RequestHandlerProvider. See below.

Next steps I think is to try to set up some tests to validate the first await issue, so we can then validate if/when we fix it.

The basic idea would be that AbstractRequestHandlerProvider checks the IRequestHandler for a property that tells it if the handler is a `write` or a `read` (or `serial` or `parallel`, or `concurrent` or `background`, whatever we settle on). It then calls a method like [this one in Blazor](https://github.com/dotnet/aspnetcore/blob/master/src/Components/Components/src/Rendering/RendererSynchronizationContextDispatcher.cs#L25) that either executes the HandleRequestAsync, or queues it up (if there is a serial request already in progress).

**Comment by dibarbet** [comment]
The NonConcurrentSynchronizationContext is seems like a no-go if it breaks on first await.

I think the razor team mentioned being interested in porting the omnisharp synchronization implementation to be usable w/out omnisharp - cc @NTaylorMullen - I don't know if we could get that done before we need to finish supporting semantic classifications though.

For the last option, is this the kind of behavior you were thinking of?
1.  All parallel requests can be executed alongside other parallel requests
2.  Any serial request could be started while previous parallel requests are in progress
3.  Subsequent parallel or serial requests must wait until the previous serial request completes

So basically, if there is a serial request in progress, all subsequent requests are queued up.  When it completes, we look at the task next on the queue - if it's a parallel one we can run it and continue taking tasks off the top without waiting for completion.  If it's a serial request, we wait for completion before taking the next task.

**Comment by NTaylorMullen** [comment]
> I think the razor team mentioned being interested in porting the omnisharp synchronization implementation to be usable w/out omnisharp - cc @NTaylorMullen - I don't know if we could get that done before we need to finish supporting semantic classifications though.

We weren't thinking about breaking O#'s synchronization system off from O# itself (it's definitely too coupled). We were considering making a VS specific O# implementation that had additional endpoints pre-defined/configured properly for VS specific LSP requests 😄 

> 2\. Any serial request could be started while previous parallel requests are in progress

You want to be careful with this one. The ideal behavior is:

1. All requests should always be run-in-order until yield
2. If the request is parallel allow the next request to run without waiting for the parallel result.
3. If the request is serial wait for the result before running any other requests.

This method ensures the most consistent approach at handling serial/parallel requests because technically whenever you await anything all bets are more-off. It also ensures that parallel request have until yield to grab the current document state which results in eventual consistency. For instance, when you do a completion request at minimum 4 things happen today:

1. textDocument/didChange to update w/ the character that triggered the completion
1. CodeActions are queried
1. text document highlight is requested
1. completion is requested.

Imagine if all of these requests started in an undefined order? If that were the case it'd be possible you're highlighting the portion of text before the textDocument/didChange happened (meaning you'd have 1 character off of your highlight).

I think the key point I want to emphasize is that every request is at least started in the order it was received

**Comment by davidwengier** [comment]
I think what I was planning is that all requests go into a queue in order. As the queue is processed, if a requests is parallel, run it without waiting. If a request is serial, wait for current running requests to finish, then run the request and wait for it to finish.

> whenever you await anything all bets are more-off

This is the bit I am concerned about, because I don't feel like I understand it deeply enough. I _think_ that NonConcurrent doesn't work for us because it never waits for a request to finish (but it will run them in order) so the only way you get the behaviour we need is having serial tasks run synchronously, and parallel tasks do an immediate `await TaskScheduler.Default;` as their first thing. Having said that, I don't understand why we couldn't enforce that behaviour by running our serial tasks in a JoinableTaskFactory.Run type scenario to force them to block.

**Comment by CyrusNajmabadi** [comment]
it feels like we could have a very simple system where we always enqueue pairs of work at a time.  Sepcifically a pair of `(serial work, parallel task that follows serial work)`.  When a new serial work is added, we enqueue one of these tuples.  Note that the parallel portion will initially be null (or empty).  Then, if parallel work is added, we just add to the parallel portion of the latest group. 

Then, we just effectively await the serial work, and a WhenAll of the parallel work.

1. no serial work runs with any other work.
2. parallel work can be as parallel as possible, until the next serial job comes in.

it seems simple and has hte semantics we want here.  It's very similar to a reader/writer lock, except with the fun constraint that readers shouldn't be able to starve new writers forever.  In this domain it's more that serial work (and new parallel work), must see the results of previous serial tasks.

**Comment by davidwengier** [comment]
I worked on this today and it ended up being much simpler than I thought. Not quite sure where the tuple you mentioned comes in @CyrusNajmabadi but I think the implementation follows the same logic that you outlined.

I put up a draft PR, would love some feedback, or some suggestions for more tests. I definitely want more tests, because the current ones pass, so obviously I don't trust them 😁
https://github.com/dotnet/roslyn/pull/46506

**Comment by CyrusNajmabadi** [comment]
ok.  i think i see waht you're saying.  there's effectively a list of mutations happening.  these must be serialized.  each produces a new snaphsot.  then, other requests can come in that do not mutate and which operate on that snapshot.  these can run concurrently but shoudlnt' affect mutation operations. 

If that's the case, this likely gets even simpler.  we simply have a queue for the mutations.  And whenever non-mutation requests come in, we just grab the result of the last mutatation and kick off the concurrent work to happen on that.

**Comment by dibarbet** [comment]
Yep I think I mainly agree with what you said, but let me first throw a wrench in :)

>  And whenever non-mutation requests come in, we just grab the result of the last mutatation and kick off the concurrent work to happen on that.

Unfortunately LSP mutations are not the only mutations that can happen.  Eventually, yes, but for now there are actions like adding/removing references that the server would hear about in a non-LSP channel.  So we can't just use the last _solution_ snapshot produced by didChange for example.

Instead the idea was to have the LSP server component keep track of the snapshots of open documents.  Then non-mutation requests would grap the snapshot results of the last mutation on open documents and apply to them workspace's solution.

**Comment by davidwengier** [comment]
I was missing that piece as being a separate task (see https://github.com/dotnet/roslyn/issues/45427#issuecomment-661554928) but it seems like they two need to be done together.

**Comment by CyrusNajmabadi** [comment]
> Unfortunately LSP mutations are not the only mutations that can happen. Eventually, yes, but for now there are actions like adding/removing references that the server would hear about in a non-LSP channel. So we can't just use the last solution snapshot produced by didChange for example.

fair enough.  i guess i mean: there is a sequence of *mutation* operations.  That includes `didChange`, but also any of these other cases you're mentioning.  Those are all serialized out and can only happen sequentially.  However, along with those, requests can come in saying: compute the results of this concurrently with whatever the 'latest mutation state' is of the system.  

Does that hold?

**Comment by davidwengier** [comment]
This is a lot simpler, yes. The algorithm would be:

```cs
var lspViewOfTheWorld = GetLspViewOfTheWorld();
while (true)
{
    var work = await _queue.DequeueAsync().ConfigureAwait(false);

    var solution = await MergeSolutions(GetSolutionSnapshot(), lspViewOfTheWorld); 
    if (work.Type == RequestProcessingMode.Serial)
    {
        await work.Callback(solution).ConfigureAwait(false);

        lspViewOfTheWorld = GetLspViewOfTheWorld();
    }
    else
    {
        // parallel work
        work.Callback(solution);
    }
}
```

The GetSolutionSnapshot() method gets the Visual Studio workspace solution. The GetLspViewOfTheWorld() gets the changes from the all of the LSP `didChange` requests. Eventually Visual Studio will catch up with those, via buffer sync, so we don't have to worry about updating the Visual Studio workspace at all. All new requests run on the latest merged snapshot, but we still need to block for mutating requests as subsequent requests need to take those mutations into account. 

Neatly, the task list is gone and we don't need to block for non-mutating requests ever.
